File: core/python/dynamic_filter_core/parser.py
"""
DSL Parser implementation for Python.
Converts string expressions into FilterTree structures.
"""

from typing import List, Optional
from enum import Enum
from dataclasses import dataclass

from .interfaces import FilterTree, Context, Condition, DSLSyntaxException

class TokenType(Enum):
    IDENTIFIER = "IDENTIFIER"
    AND = "AND"
    OR = "OR"
    NOT = "NOT"
    LEFT_PAREN = "LEFT_PAREN"
    RIGHT_PAREN = "RIGHT_PAREN"

@dataclass
class Token:
    type: TokenType
    value: str

class DSLParser:
    """Implementation of the DSL parser that converts string expressions into FilterTree structures."""
    
    def parse(self, dsl_expression: str) -> FilterTree:
        """Parses a DSL expression string into a FilterTree."""
        if not dsl_expression or not dsl_expression.strip():
            raise DSLSyntaxException("DSL expression cannot be null or empty")
        
        trimmed = dsl_expression.strip()
        return self._parse_expression(trimmed)
    
    def _parse_expression(self, expression: str) -> FilterTree:
        """Parses an expression into tokens and builds a FilterTree."""
        tokens = self._tokenize(expression)
        return self._parse_tokens(tokens)
    
    def _tokenize(self, expression: str) -> List[Token]:
        """Tokenizes the expression string."""
        tokens = []
        current_token = ""
        
        for i, c in enumerate(expression):
            if c.isspace():
                # End current token if any
                if current_token:
                    tokens.append(Token(TokenType.IDENTIFIER, current_token))
                    current_token = ""
                continue
            
            if c in "&|!()":
                # End current token if any
                if current_token:
                    tokens.append(Token(TokenType.IDENTIFIER, current_token))
                    current_token = ""
                
                # Add operator/parenthesis token
                token_type = self._get_token_type(c)
                tokens.append(Token(token_type, c))
            elif c.isalnum() or c == '_':
                current_token += c
            else:
                raise DSLSyntaxException(f"Invalid character '{c}' at position {i}")
        
        # Add final token if any
        if current_token:
            tokens.append(Token(TokenType.IDENTIFIER, current_token))
        
        return tokens
    
    def _get_token_type(self, c: str) -> TokenType:
        """Gets the token type for a character."""
        token_map = {
            '&': TokenType.AND,
            '|': TokenType.OR,
            '!': TokenType.NOT,
            '(': TokenType.LEFT_PAREN,
            ')': TokenType.RIGHT_PAREN
        }
        if c not in token_map:
            raise ValueError(f"Unknown operator: {c}")
        return token_map[c]
    
    def _parse_tokens(self, tokens: List[Token]) -> FilterTree:
        """Parses tokens into a FilterTree using the Shunting Yard algorithm."""
        if not tokens:
            raise DSLSyntaxException("Empty expression")
        
        # Convert to postfix notation using Shunting Yard algorithm
        postfix = self._infix_to_postfix(tokens)
        
        # Build expression tree from postfix notation
        return self._build_expression_tree(postfix)
    
    def _infix_to_postfix(self, tokens: List[Token]) -> List[Token]:
        """Converts infix notation to postfix notation using Shunting Yard algorithm."""
        output = []
        operators = []
        
        for token in tokens:
            if token.type == TokenType.IDENTIFIER:
                output.append(token)
            elif token.type == TokenType.NOT:
                operators.append(token)
            elif token.type in (TokenType.AND, TokenType.OR):
                while (operators and 
                       operators[-1].type != TokenType.LEFT_PAREN and
                       self._get_precedence(operators[-1].type) >= self._get_precedence(token.type)):
                    output.append(operators.pop())
                operators.append(token)
            elif token.type == TokenType.LEFT_PAREN:
                operators.append(token)
            elif token.type == TokenType.RIGHT_PAREN:
                while operators and operators[-1].type != TokenType.LEFT_PAREN:
                    output.append(operators.pop())
                if not operators:
                    raise DSLSyntaxException("Mismatched parentheses")
                operators.pop()  # Remove left parenthesis
        
        while operators:
            op = operators.pop()
            if op.type == TokenType.LEFT_PAREN:
                raise DSLSyntaxException("Mismatched parentheses")
            output.append(op)
        
        return output
    
    def _get_precedence(self, token_type: TokenType) -> int:
        """Gets the precedence of a token type."""
        precedence_map = {
            TokenType.NOT: 3,
            TokenType.AND: 2,
            TokenType.OR: 1
        }
        return precedence_map.get(token_type, 0)
    
    def _build_expression_tree(self, postfix: List[Token]) -> FilterTree:
        """Builds an expression tree from postfix notation."""
        stack = []
        
        for token in postfix:
            if token.type == TokenType.IDENTIFIER:
                stack.append(IdentifierNode(token.value))
            elif token.type == TokenType.NOT:
                if not stack:
                    raise DSLSyntaxException("Invalid expression: NOT operator without operand")
                stack.append(NotNode(stack.pop()))
            elif token.type == TokenType.AND:
                if len(stack) < 2:
                    raise DSLSyntaxException("Invalid expression: AND operator requires two operands")
                right = stack.pop()
                left = stack.pop()
                stack.append(AndNode(left, right))
            elif token.type == TokenType.OR:
                if len(stack) < 2:
                    raise DSLSyntaxException("Invalid expression: OR operator requires two operands")
                right = stack.pop()
                left = stack.pop()
                stack.append(OrNode(left, right))
        
        if len(stack) != 1:
            raise DSLSyntaxException("Invalid expression: malformed syntax")
        
        return stack[0]

# Node classes for the expression tree
class IdentifierNode(FilterTree):
    def __init__(self, identifier: str):
        self.identifier = identifier
    
    def generate(self, context: Context) -> Condition:
        condition = context.get_condition(self.identifier)
        if not condition:
            raise ValueError(f"No condition found for identifier: {self.identifier}")
        return condition

class NotNode(FilterTree):
    def __init__(self, operand: FilterTree):
        self.operand = operand
    
    def generate(self, context: Context) -> Condition:
        return self.operand.generate(context).not_()

class AndNode(FilterTree):
    def __init__(self, left: FilterTree, right: FilterTree):
        self.left = left
        self.right = right
    
    def generate(self, context: Context) -> Condition:
        return self.left.generate(context).and_(self.right.generate(context))

class OrNode(FilterTree):
    def __init__(self, left: FilterTree, right: FilterTree):
        self.left = left
        self.right = right
    
    def generate(self, context: Context) -> Condition:
        return self.left.generate(context).or_(self.right.generate(context))

File: core/typescript/src/parser.ts
/**
 * DSL Parser implementation for TypeScript/JavaScript.
 * Converts string expressions into FilterTree structures.
 */

import { FilterTree, Context, Condition, DSLSyntaxException } from './interfaces';

export enum TokenType {
  IDENTIFIER = 'IDENTIFIER',
  AND = 'AND',
  OR = 'OR',
  NOT = 'NOT',
  LEFT_PAREN = 'LEFT_PAREN',
  RIGHT_PAREN = 'RIGHT_PAREN'
}

export interface Token {
  type: TokenType;
  value: string;
}

export class DSLParser {
  parse(dslExpression: string): FilterTree {
    if (!dslExpression || dslExpression.trim().length === 0) {
      throw new DSLSyntaxException('DSL expression cannot be null or empty');
    }

    const trimmed = dslExpression.trim();
    return this.parseExpression(trimmed);
  }

  private parseExpression(expression: string): FilterTree {
    const tokens = this.tokenize(expression);
    return this.parseTokens(tokens);
  }

  private tokenize(expression: string): Token[] {
    const tokens: Token[] = [];
    let currentToken = '';

    for (let i = 0; i < expression.length; i++) {
      const c = expression[i];

      if (/\s/.test(c)) {
        // End current token if any
        if (currentToken.length > 0) {
          tokens.push({ type: TokenType.IDENTIFIER, value: currentToken });
          currentToken = '';
        }
        continue;
      }

      if (c === '&' || c === '|' || c === '!' || c === '(' || c === ')') {
        // End current token if any
        if (currentToken.length > 0) {
          tokens.push({ type: TokenType.IDENTIFIER, value: currentToken });
          currentToken = '';
        }

        // Add operator/parenthesis token
        const type = this.getTokenType(c);
        tokens.push({ type, value: c });
      } else if (/[a-zA-Z0-9_]/.test(c)) {
        currentToken += c;
      } else {
        throw new DSLSyntaxException(`Invalid character '${c}' at position ${i}`);
      }
    }

    // Add final token if any
    if (currentToken.length > 0) {
      tokens.push({ type: TokenType.IDENTIFIER, value: currentToken });
    }

    return tokens;
  }

  private getTokenType(c: string): TokenType {
    switch (c) {
      case '&': return TokenType.AND;
      case '|': return TokenType.OR;
      case '!': return TokenType.NOT;
      case '(': return TokenType.LEFT_PAREN;
      case ')': return TokenType.RIGHT_PAREN;
      default: throw new Error(`Unknown operator: ${c}`);
    }
  }

  private parseTokens(tokens: Token[]): FilterTree {
    if (tokens.length === 0) {
      throw new DSLSyntaxException('Empty expression');
    }

    // Convert to postfix notation using Shunting Yard algorithm
    const postfix = this.infixToPostfix(tokens);

    // Build expression tree from postfix notation
    return this.buildExpressionTree(postfix);
  }

  private infixToPostfix(tokens: Token[]): Token[] {
    const output: Token[] = [];
    const operators: Token[] = [];

    for (const token of tokens) {
      switch (token.type) {
        case TokenType.IDENTIFIER:
          output.push(token);
          break;
        case TokenType.NOT:
          operators.push(token);
          break;
        case TokenType.AND:
        case TokenType.OR:
          while (operators.length > 0 &&
                 operators[operators.length - 1].type !== TokenType.LEFT_PAREN &&
                 this.getPrecedence(operators[operators.length - 1].type) >= this.getPrecedence(token.type)) {
            output.push(operators.pop()!);
          }
          operators.push(token);
          break;
        case TokenType.LEFT_PAREN:
          operators.push(token);
          break;
        case TokenType.RIGHT_PAREN:
          while (operators.length > 0 && operators[operators.length - 1].type !== TokenType.LEFT_PAREN) {
            output.push(operators.pop()!);
          }
          if (operators.length === 0) {
            throw new DSLSyntaxException('Mismatched parentheses');
          }
          operators.pop(); // Remove left parenthesis
          break;
      }
    }

    while (operators.length > 0) {
      const op = operators.pop()!;
      if (op.type === TokenType.LEFT_PAREN) {
        throw new DSLSyntaxException('Mismatched parentheses');
      }
      output.push(op);
    }

    return output;
  }

  private getPrecedence(type: TokenType): number {
    switch (type) {
      case TokenType.NOT: return 3;
      case TokenType.AND: return 2;
      case TokenType.OR: return 1;
      default: return 0;
    }
  }

  private buildExpressionTree(postfix: Token[]): FilterTree {
    const stack: FilterTree[] = [];

    for (const token of postfix) {
      switch (token.type) {
        case TokenType.IDENTIFIER:
          stack.push(new IdentifierNode(token.value));
          break;
        case TokenType.NOT:
          if (stack.length === 0) {
            throw new DSLSyntaxException('Invalid expression: NOT operator without operand');
          }
          stack.push(new NotNode(stack.pop()!));
          break;
        case TokenType.AND:
          if (stack.length < 2) {
            throw new DSLSyntaxException('Invalid expression: AND operator requires two operands');
          }
          const rightAnd = stack.pop()!;
          const leftAnd = stack.pop()!;
          stack.push(new AndNode(leftAnd, rightAnd));
          break;
        case TokenType.OR:
          if (stack.length < 2) {
            throw new DSLSyntaxException('Invalid expression: OR operator requires two operands');
          }
          const rightOr = stack.pop()!;
          const leftOr = stack.pop()!;
          stack.push(new OrNode(leftOr, rightOr));
          break;
      }
    }

    if (stack.length !== 1) {
      throw new DSLSyntaxException('Invalid expression: malformed syntax');
    }

    return stack[0];
  }
}

// Node classes for the expression tree
class IdentifierNode implements FilterTree {
  constructor(private identifier: string) {}

  generate(context: Context): Condition {
    const condition = context.getCondition(this.identifier);
    if (!condition) {
      throw new Error(`No condition found for identifier: ${this.identifier}`);
    }
    return condition;
  }
}

class NotNode implements FilterTree {
  constructor(private operand: FilterTree) {}

  generate(context: Context): Condition {
    return this.operand.generate(context).not();
  }
}

class AndNode implements FilterTree {
  constructor(private left: FilterTree, private right: FilterTree) {}

  generate(context: Context): Condition {
    return this.left.generate(context).and(this.right.generate(context));
  }
}

class OrNode implements FilterTree {
  constructor(private left: FilterTree, private right: FilterTree) {}

  generate(context: Context): Condition {
    return this.left.generate(context).or(this.right.generate(context));
  }
}
